### PR TITLE
IPMPROG-1105 - fix

### PR DESCRIPTION
**Pull request by jeffdegott**
* IPMPROG-1105-fix
Signed-off-by: Degott, Francois Regis <FrancoisRegisDegott@eaton.com>

### CODE REVIEW

File: src/list-in.cpp [hunk 1]
```diff
@@ -136,66 +136,75 @@ struct Outlet
...
+static std::pair<std::string, std::string> getOutletNumberAndProperty(const std::string& extAttributeName)
 {
-    auto        dot1    = extAttributeName.find_first_of(".");
-    std::string oNumber = extAttributeName.substr(dot1 + 1);
-    auto        dot2    = oNumber.find_first_of(".");
-    oNumber             = oNumber.substr(0, dot2);
-    return oNumber;
+    std::pair<std::string, std::string> result; // empty
+
+    if (extAttributeName.find("outlet.") != 0)
```
Comment (perrettecl) [inline src/list-in.cpp:150]: If should always have {} to avoid issues

File: src/list-in.cpp [hunk 2]
```diff
@@ -136,66 +136,75 @@ struct Outlet
...
+
+    result.first = aux.substr(0, dot2); // number
+    result.second = aux.substr(dot2 + 1); // property name
+    return result;
 }
 
-static std::map<std::string, Outlet> collectOutlets(const db::asset::Attributes& ext)
+// returns a ref. to existing Outlet entry in outlets
+
+static Outlet* findOrCreateOutlet(const std::string& key, std::map<std::string, Outlet>& outlets)
```
Comment (perrettecl) [inline src/list-in.cpp:167]: This function is useless: operator[] does already this see https://en.cppreference.com/w/cpp/container/map/operator_at

Running this code is a proof of it:
`
#include <iostream>
#include <string>
#include <map>
 
int main()
{
    std::map<int, std::string>  myMap;
    myMap[0] = "data0";
    std::cout << "Map at 0: '" << myMap[0] <<"'"<< std::endl;
    
    auto & element0 = myMap[0];
    element0 = "new data";
    std::cout << "Map at 0: '" << element0 <<"'"<< std::endl;
    
    auto & element1 = myMap[1];
    std::cout << "Map at 1: '" << element1 <<"'"<< std::endl;
    element1 = "data1";
    std::cout << "Map at 1: '" << element1 <<"'"<< std::endl;
}

`

Output:

> Map at 0: 'data0'
> Map at 0: 'new data'
> Map at 1: ''
> Map at 1: 'data1'

File: src/list-in.cpp [hunk 3]
```diff
@@ -136,66 +136,75 @@ struct Outlet
...
+{
     std::map<std::string, Outlet> outlets;
 
     for (const auto& [key, value] : ext) {
-        if (key.find("outlet.") != 0) {
-            continue;
+        // key match "outlet.<number>.<property-name>"?
+        // pair.first as number, .second as property-name
+        auto pair = getOutletNumberAndProperty(key);
+        if (pair.first.empty() || pair.second.empty())
```
Comment (perrettecl) [inline src/list-in.cpp:183]: if without {}

File: src/list-in.cpp [hunk 4]
```diff
@@ -136,66 +136,75 @@ struct Outlet
...
+            continue; // not an 'outlet' ext. attr
+
+        if (pair.first == "0")
```
Comment (perrettecl) [inline src/list-in.cpp:186]: if without {}

File: src/list-in.cpp [hunk 5]
```diff
@@ -136,66 +136,75 @@ struct Outlet
...
     db::asset::ExtAttrValue group;
+    db::asset::ExtAttrValue name;
+    db::asset::ExtAttrValue switchable;
 };
 
-static std::string getOutletNumber(const std::string& extAttributeName)
+// parse ext. attribute as "outlet.<number>.<property-name>
+// returns <outlet number, property name> (empty if inconsistent)
+
+static std::pair<std::string, std::string> getOutletNumberAndProperty(const std::string& extAttributeName)
```
Comment (etnKlendathu) [inline src/list-in.cpp:146]:
```suggestion
static fty::Expected<std::pair<std::string, std::string>> getOutletNumberAndProperty(const std::string& extAttributeName)
```

File: src/list-in.cpp (hunk 2 above)
Comment (etnKlendathu) [inline src/list-in.cpp:167]: This function is potentially dangerous. If you clean map or something else your pointer become invalid...

**Review by eric3873 — APPROVED** [review]

**Review by perrettecl — APPROVED** [review]

**Pull request merged**